qiling/extensions/afl/qlfuzzer.py

```diff
@@ -0,0 +1,157 @@
+
+        Args:
+            crashes: executable addresses to hook
+        """
+
+        def __crash(ql: Qiling) -> None:
+            os.abort()
+
+        for address in crashes:
+            self.ql.hook_address(__crash, address)
+
+    def __install_kickoff_hook(self, infilename: str, entry: int, exits: Collection[int]) -> None:
+        def __kickoff(ql: Qiling):
+            """Have Unicorn forked and start instrumentation.
+            """
+
+            # this is just a one-time hook; remove it
+            ko_hook.remove()
+
+            afl.ql_afl_fuzz(ql, infilename, self.feed_input, exits)
+
+        # set afl instrumentation [re]starting point
+        ko_hook = self.ql.hook_address(__kickoff, entry)
+
+    def stage_call_site(self, params: Sequence[int]) -> None:
+        """Stage parameters for a function call.
+        This method provides a convinient way to set up parameters when fuzzing a function call.
+
+        Args:
+            params: a sequence of integer values to set as parameters
+        """
+
+        self.ql.os.fcall.writeParams([(POINTER, p) for p in params])
+
+    @abstractmethod
+    def feed_input(self, ql: Qiling, stimuli: bytes, pround: int) -> bool:
+        """A callback method invoked by AFL whenever a new fuzzing stimuli is generated.
+        The method may manipulate the stimuli to its needs or use it as-is, and ultimately
+        responsible to place it where the fuzzed program expects its input to be found, e.g.:
+        stdin, file, socket, memory, etc.
+
+        Args:
+            ql: qiling instance
+            stimuli: newly generated input to the fuzzed program
+            pround: iteration number within a persistent session. if persistency was not set,
+            round value is expected to be 0 every time
+
+        Returns: a boolean indicator of whether AFL should proceed with this fuzzing iteration
+        or not (i.e. in case the generated stimuli does not satisfy fuzzing logic criteria)
+        """
+
+    def ea(self, offset: int, module: Optional[str] = None, *, casefold: bool = False) -> int:
+        """Get the effective address of a file offset.
+
+        Args:
+            offset: file offset
+            module: module basename (the emulated binary, by default)
+            casefold: match module name case-insensitively. this becomes useful when windows
+            binaries load their libraries using arbitrary case names
+
+        Returns: the effective address of `offset` using `module` base address.
+        Raises: `KeyError` if the requested module was not loaded
+        """
+
+        image = self.ql.loader.get_image_by_name(module or os.path.basename(self.ql.argv[0]), casefold=casefold)
+
+        if image is None:
+            raise KeyError(f'could not find a loaded module named "{module}"')
+
+        return image.base + offset
+
+    def setup(self, infilename: str, entry: int, exits: Collection[int], crashes: Optional[Collection[int]] = None) -> None:
```

Comment:
It would be better if `infile` can be multiple files though our cases always use single one.

Reply:
How would you suggest to do that?
The `uc_afl_fuzz_custom` API in `unicornafl` takes just one filename.